crates/ruff_linter/src/rules/flake8_cognitive_complexity/rules/function_is_too_cognitive_complex.rs

```diff
@@ -0,0 +1,281 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, identifier::Identifier, Expr, Stmt};
+
+/// ## What it does
+/// Checks for functions with a high cognitive complexity.
+///
+/// ## Why is this bad?
+/// Cognitive complexity is a metric based of a whitepaper by G. Ann Campbell.
+///
+///  > Cognitive Complexity breaks from the practice of using mathematical models to
+///  > assess software maintainability. It starts from the precedents set by Cyclomatic
+///  > Complexity, but uses human judgment to assess how structures should be
+///  > counted, and to decide what should be added to the model as a whole. As a
+///  > result, it yields method complexity scores which strike programmers as fairer
+///  > relative assessments of understandability than have been available with previous
+///  > models. Further, because Cognitive Complexity charges no "cost of entry" for
+///  > a method, it produces those fairer relative assessments not just at the method
+///  > level, but also at the class and application levels.
+///
+/// ref [{Cognitive Complexity} a new way of measuring understandability](file:///home/anderss/Downloads/Cognitive_Complexity_Sonar_Guide_2023%20(1).pdf).
```

Comment:
This links to a local file. Is this available online too and we could link to it instead?

Reply:
Oops, a blunder, yes, it is the same `.pdf` that I used.